fasm/__init__.py

```diff
@@ -46,122 +60,134 @@ class ValueFormat(enum.Enum):
-        value_format = ValueFormat.VERILOG_HEX
-    elif verilog_value.binary_value:
-        value = int(verilog_value.binary_value.replace('_', ''), 2)
-        value_format = ValueFormat.VERILOG_BINARY
-    elif verilog_value.decimal_value:
-        value = int(verilog_value.decimal_value.replace('_', ''), 10)
-        value_format = ValueFormat.VERILOG_DECIMAL
-    elif verilog_value.octal_value:
-        value = int(verilog_value.octal_value.replace('_', ''), 8)
-        value_format = ValueFormat.VERILOG_OCTAL
+# The following functions decode the binary format generated by parse_fasm.
+# This is a lightweight binary format designed to be fast to decode.
+
+# Match a tag and retrieve length from the header.
+def get_header(tags, data, i):
+    if not data[i] in bytes(tags, 'ascii'):
+        # print('tag mismatch: got {}, expecting {} at position {}'.format(data[i], tags, i))
+        # print('data: {}', data[:i])
+        return None, i
+
+    return int.from_bytes(data[i+1:i+5], byteorder), i+5
+
+# Decode a tagged string
+def tagged_string_from_bytes(tag, data, i):
+    length, i = get_header(tag, data, i)
+    if length is None:
+        return None, i
+
+    return data[i:i+length].decode('ascii'), i+length
+
+# Decode a FASM address: [x:y]
+def fasm_address_from_bytes(data, i):
+    length, i = get_header(':', data, i)
+    if not length:
+        return None, None, i
+
+    end = int.from_bytes(data[i:i+4], byteorder)
+    start = int.from_bytes(data[i+4:i+8], byteorder) if length == 8 else None
+    return start, end, i+length
+
+# Decode a FASM value
+def fasm_value_from_bytes(data, i):
+    tag = data[i]
+    if tag == b'p':
+        length = 4
+        i = i + 1
     else:
-        assert False, verilog_value
-
-    if width is not None:
-        assert_max_width(width, value)
+        length, i = get_header('hobd', data, i)
 
-    return width, value, value_format
+    if not length:
+        return None, None, i
 
+    return int.from_bytes(data[i:i+length], byteorder), tag_to_value_format[tag], i+length
 
-def set_feature_model_to_tuple(set_feature_model):
-    start = None
-    end = None
-    value = 1
-    address_width = 1
-    value_format = None
+# Decode a set feature: feature = value
+def fasm_set_feature_from_bytes(data, i):
+    length, i = get_header('s', data, i)
+    if not length:
+        return None, i
 
-    if set_feature_model.feature_address:
-        if set_feature_model.feature_address.address2:
-            end = int(set_feature_model.feature_address.address1, 10)
-            start = int(set_feature_model.feature_address.address2, 10)
-            address_width = end - start + 1
-        else:
-            start = int(set_feature_model.feature_address.address1, 10)
-            end = None
-            address_width = 1
+    feature, p = tagged_string_from_bytes('f', data, i)
+    start, end, p = fasm_address_from_bytes(data, p)
+    value, value_format, p = fasm_value_from_bytes(data, p)
```

Comment:
You are missing the old format assertions about value versus address matching?

Reply:
I've added the assertion back. Caught a bug! Thanks.